### PR TITLE
StripePI: Enable Apple Pay and Google Pay payment methods

### PR DESCRIPTION
**Pull request by gasb150**
Summary:
---------------------------------------
In order to add  Apple Pay and Google Pay support to Stripe Payment Intents
this commit enables the option to store those payment methods for use in the Payment Intent creation.

Local Tests:
---------------------------------------
35 tests, 183 assertions, 0 failures, 0 errors, 0 pendings, 0 omissions, 0 notifications
100% passed

Remote Tests:
---------------------------------------
Finished in 168.186212 seconds.
68 tests, 317 assertions, 0 failures, 0 errors, 0 pendings, 0 omissions, 0 notifications
100% passed

RuboCop:
---------------------------------------
725 files inspected, no offenses detected

### CODE REVIEW

File: test/unit/gateways/stripe_payment_intents_test.rb [hunk 1]
```diff
@@ -26,12 +26,15 @@ def setup
 
     @apple_pay = apple_pay_payment_token
     @google_pay = network_tokenization_credit_card(
-      '4777777777777778',
-      payment_cryptogram: 'BwAQCFVQdwEAABNZI1B3EGLyGC8=',
-      verification_value: '987',
-      source: :google_pay,
+      '4242424242424242',
+      payment_cryptogram: 'reddelicious',
+      source: :android_pay,
```
Comment (curiousepic) [inline test/unit/gateways/stripe_payment_intents_test.rb:31]: Why was this changed? Is it more accurate test data?
Are we sure there are no differences between the source being `android_pay` and `google_pay`?

File: test/remote/gateways/remote_stripe_payment_intents_test.rb [hunk 2]
```diff
@@ -64,6 +86,14 @@ def test_successful_purchase
...
   end
 
+  def test_successful_google_pay_store
+    options = {
+      currency: 'GBP'
+    }
+    assert store = @gateway.store(@google_pay, options)
+    assert store.params['id'].start_with?('cus_')
+  end
+
```
Comment (curiousepic) [inline test/remote/gateways/remote_stripe_payment_intents_test.rb:96]: Let's add one for Apple Pay if we can.

File: test/remote/gateways/remote_stripe_payment_intents_test.rb [hunk 3]
```diff
@@ -35,8 +35,30 @@ def setup
...
+      payment_cryptogram: 'reddelicious',
+      source: 'apple_pay',
+      brand: 'visa',
+      eci: '05',
+      month: '09',
+      year: '2030',
+      first_name: 'Longbob',
+      last_name: 'Longsen'
+    )
+    @credit_card_token = 'card_1KBQpkAWOtgoysogjIqSmzES'
```
Comment (curiousepic) [inline test/remote/gateways/remote_stripe_payment_intents_test.rb:61]: Let's not include this if it is not used

File: lib/active_merchant/billing/gateways/stripe_payment_intents.rb [hunk 4]
```diff
@@ -189,7 +189,12 @@ def store(payment_method, options = {})
 
         # If customer option is provided, create a payment method and attach to customer id
         # Otherwise, create a customer, then attach
-        if payment_method.is_a?(StripePaymentToken) || payment_method.is_a?(ActiveMerchant::Billing::CreditCard)
+        card_source_pay = payment_method.source.to_s if defined?(payment_method.source)
+        card_brand_pay = card_brand(payment_method) unless payment_method.is_a?(String) || payment_method.nil?
+        if NO_WALLET_SUPPORT.include?(card_source_pay) || NO_WALLET_SUPPORT.include?(card_brand_pay)
+          result = super(payment_method, options)
```
Comment (curiousepic) [inline lib/active_merchant/billing/gateways/stripe_payment_intents.rb:195]: I'm a little confused about why we need to call to super and add logic to the base Stripe adapter here. Is it just a matter of not duplicating code?

File: lib/active_merchant/billing/gateways/stripe.rb [hunk 5]
```diff
@@ -205,6 +205,11 @@ def store(payment, options = {})
           return bank_token_response unless bank_token_response.success?
 
           params = { source: bank_token_response.params['token']['id'] }
+        elsif payment.is_a?(ActiveMerchant::Billing::NetworkTokenizationCreditCard)
+          tokenize_apple_google = tokenize_apple_google(payment, options)
+          return tokenize_apple_google unless tokenize_apple_google.success?
+
+          params = { source: tokenize_apple_google.params['token']['id'] }
```
Comment (curiousepic) [inline lib/active_merchant/billing/gateways/stripe.rb:212]: Would these changes also enable support for storing AP/GP via the base Stripe gateway?

**Comment by curiousepic** [comment]
This looks pretty good - I left a few questions, but to summarize them: It's a little weird that supporting this on PI requires changes to the Base gateway, which already kind of supports AP/GP (but maybe not storing them)? So I'm wondering if we can also support storing them on the Base gateway as part of this update...

File: lib/active_merchant/billing/gateways/stripe.rb (hunk 5 above)
Comment (gasb150) [reply lib/active_merchant/billing/gateways/stripe.rb:212]: yes

File: lib/active_merchant/billing/gateways/stripe.rb [hunk 6]
```diff
@@ -262,6 +267,32 @@ def tokenize_apple_pay_token(apple_pay_payment_token, options = {})
         end
       end
 
+      def tokenize_apple_google(payment, options = {})
+        if payment.inspect.include?('google_pay')
```
Comment (curiousepic) [inline lib/active_merchant/billing/gateways/stripe.rb:271]: I would check `payment.source == :google_pay` directly here, rather than `.include?`

There is also some possibility the `source` will be something other than these two (`android_pay` or `network_token`), which we should handle/guard against with a guard clause or a third case.

File: lib/active_merchant/billing/gateways/stripe_payment_intents.rb [hunk 7]
```diff
@@ -189,7 +189,12 @@ def store(payment_method, options = {})
 
         # If customer option is provided, create a payment method and attach to customer id
         # Otherwise, create a customer, then attach
-        if payment_method.is_a?(StripePaymentToken) || payment_method.is_a?(ActiveMerchant::Billing::CreditCard)
+        card_source_pay = payment_method.source.to_s if defined?(payment_method.source)
+        card_brand_pay = card_brand(payment_method) unless payment_method.is_a?(String) || payment_method.nil?
```
Comment (curiousepic) [inline lib/active_merchant/billing/gateways/stripe_payment_intents.rb:193]: I don't think this final `payment_method.nil?` check is necessary at this point...

File: lib/active_merchant/billing/gateways/stripe_payment_intents.rb [hunk 8]
```diff
@@ -189,7 +189,12 @@ def store(payment_method, options = {})
...
+        if NO_WALLET_SUPPORT.include?(card_source_pay) || NO_WALLET_SUPPORT.include?(card_brand_pay)
+          result = super(payment_method, options)
+          return result if result.is_a?(ActiveMerchant::Billing::Response)
```
Comment (curiousepic) [inline lib/active_merchant/billing/gateways/stripe_payment_intents.rb:196]: What are the other potential outcomes here other than an AM Response? And what happens if it's not?

**Comment by gasb150** [comment]
We have decided to modify the current implementation a little bit. Therefore we will close this pull request to create a new one. Thanks so much.
